src/Interpreters/ConcurrentHashJoin.cpp

```diff
@@ -458,4 +612,59 @@ UInt64 calculateCacheKey(
 
     return hash.get64();
 }
+
+void ConcurrentHashJoin::onBuildPhaseFinish()
+{
+    for (auto & hash_join : hash_joins)
+    {
+        // It cannot be called concurrently with other IJoin methods
+        hash_join->data->onBuildPhaseFinish();
+    }
+
+    if (!hash_joins[0]->data->twoLevelMapIsUsed())
+        return;
+
+    // At this point, the build phase is finished. We need to build a shared common hash map to be used in the probe phase.
+    // It is done in two steps:
+    //     1. Merge hash maps into a single one. For that, we iterate over all sub-maps and move buckets from the current `HashJoin` instance to the common map.
```

Comment:
So, does the resulting two-level map have `<number of HashJoin instances> x <number of levels in each map in HashJoin>` maps? How is it different from building a single two-level map?